cpp/src/io/parquet/experimental/page_index_filter_utils.cu

```diff
@@ -0,0 +1,253 @@
+  // Compute total number of row groups
+  auto const total_row_groups =
+    std::accumulate(row_group_indices.begin(),
+                    row_group_indices.end(),
+                    size_t{0},
+                    [](auto sum, auto const& rg_indices) { return sum + rg_indices.size(); });
+
+  std::vector<size_type> page_row_offsets;
+  page_row_offsets.push_back(0);
+  size_type max_page_size = 0;
+
+  std::for_each(thrust::counting_iterator<size_t>(0),
+                thrust::counting_iterator(row_group_indices.size()),
+                [&](auto const src_idx) {
+                  // For all row groups in this source
+                  auto const& rg_indices = row_group_indices[src_idx];
+                  std::optional<size_type> colchunk_iter_offset{};
+                  std::for_each(rg_indices.begin(), rg_indices.end(), [&](auto const& rg_idx) {
+                    auto const& row_group = per_file_metadata[src_idx].row_groups[rg_idx];
+                    // Find the column chunk with the given schema index
+                    if (not colchunk_iter_offset.has_value() or
+                        row_group.columns[colchunk_iter_offset.value()].schema_idx != schema_idx) {
+                      colchunk_iter_offset = find_colchunk_iter_offset(row_group, schema_idx);
+                    }
+                    auto const& colchunk_iter =
+                      row_group.columns.begin() + colchunk_iter_offset.value();
+                    auto const& offset_index       = colchunk_iter->offset_index.value();
+                    auto const row_group_num_pages = offset_index.page_locations.size();
+                    std::for_each(thrust::counting_iterator<size_t>(0),
+                                  thrust::counting_iterator(row_group_num_pages),
+                                  [&](auto const page_idx) {
+                                    int64_t const first_row_idx =
+                                      offset_index.page_locations[page_idx].first_row_index;
+                                    int64_t const last_row_idx =
+                                      (page_idx < row_group_num_pages - 1)
+                                        ? offset_index.page_locations[page_idx + 1].first_row_index
+                                        : row_group.num_rows;
+                                    auto const page_size = last_row_idx - first_row_idx;
+                                    max_page_size = std::max<size_type>(max_page_size, page_size);
+                                    page_row_offsets.push_back(page_row_offsets.back() + page_size);
+                                  });
+                  });
+                });
+
+  return {std::move(page_row_offsets), max_page_size};
+}
+
+rmm::device_uvector<size_type> compute_page_indices_async(
+  cudf::host_span<cudf::size_type const> page_row_counts,
+  cudf::host_span<cudf::size_type const> page_row_offsets,
+  cudf::size_type total_rows,
+  rmm::cuda_stream_view stream)
```

Comment:
https://github.com/rapidsai/cudf/blob/main/cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md#output-memory

Any libcudf API that allocates memory that is returned to a user must accept a `rmm::device_async_resource_ref` as the last parameter.